### ISSUE TITLE
Unable to enable maintenance mode when env.php is read only

### ISSUE THREAD

**Issue by ihor-sviziev** [magento/magento2#24229]
### Preconditions (*)
1. Magento 2.3.2 & 2.3-develop
2. Make env.php file read only: `chmod -w app/etc/env.php` or `chmod 754 app/etc/env.php`

### Steps to reproduce (*)
1. Run `php bin/magento maintenance:enable -v`

### Expected result (*)
1. Maintenance mode should be enabled

### Actual result (*)
1. Maintenance mode is not enabled
2. Command fails with following output
![image](https://user-images.githubusercontent.com/1873745/63500031-baa38b80-c4d1-11e9-87d1-a20a3845d8ed.png)

```
In PathValidator.php line 63:
                                                                                                          
  [Magento\Framework\Exception\ValidatorException]                                                        
  Path "/var/www/html/magento2/app/etc/" cannot be used with directory "/var/www/html/magento2/app/etc/"  
                                                                                                          

Exception trace:
 () at /var/www/html/magento2/vendor/magento/framework/Filesystem/Directory/PathValidator.php:63
 Magento\Framework\Filesystem\Directory\PathValidator->validate() at /var/www/html/magento2/vendor/magento/framework/Filesystem/Directory/Read.php:82
 Magento\Framework\Filesystem\Directory\Read->validatePath() at /var/www/html/magento2/vendor/magento/framework/Filesystem/Directory/Read.php:111
 Magento\Framework\Filesystem\Directory\Read->getAbsolutePath() at /var/www/html/magento2/vendor/magento/framework/Filesystem/Directory/Write.php:57
 Magento\Framework\Filesystem\Directory\Write->assertWritable() at /var/www/html/magento2/vendor/magento/framework/Filesystem/Directory/Write.php:326
 Magento\Framework\Filesystem\Directory\Write->openFile() at /var/www/html/magento2/vendor/magento/framework/Filesystem/Directory/Write.php:343
 Magento\Framework\Filesystem\Directory\Write->writeFile() at /var/www/html/magento2/vendor/magento/framework/App/DeploymentConfig/Writer.php:146
 Magento\Framework\App\DeploymentConfig\Writer->saveConfig() at /var/www/html/magento2/vendor/magento/framework/App/Cache/State.php:104
 Magento\Framework\App\Cache\State->persist() at /var/www/html/magento2/vendor/magento/framework/App/Cache/Manager.php:77
 Magento\Framework\App\Cache\Manager->setEnabled() at /var/www/html/magento2/vendor/magento/module-page-cache/Observer/SwitchPageCacheOnMaintenance.php:73
 Magento\PageCache\Observer\SwitchPageCacheOnMaintenance->turnOffFullPageCache() at /var/www/html/magento2/vendor/magento/module-page-cache/Observer/SwitchPageCacheOnMaintenance.php:55
 Magento\PageCache\Observer\SwitchPageCacheOnMaintenance->execute() at /var/www/html/magento2/vendor/magento/framework/Event/Invoker/InvokerDefault.php:88
 Magento\Framework\Event\Invoker\InvokerDefault->_callObserverMethod() at /var/www/html/magento2/vendor/magento/framework/Event/Invoker/InvokerDefault.php:74
 Magento\Framework\Event\Invoker\InvokerDefault->dispatch() at /var/www/html/magento2/vendor/magento/framework/Event/Manager.php:66
 Magento\Framework\Event\Manager->dispatch() at /var/www/html/magento2/vendor/magento/framework/App/MaintenanceMode.php:82
 Magento\Framework\App\MaintenanceMode->set() at /var/www/html/magento2/setup/src/Magento/Setup/Console/Command/AbstractMaintenanceCommand.php:90
 Magento\Setup\Console\Command\AbstractMaintenanceCommand->execute() at /var/www/html/magento2/vendor/symfony/console/Command/Command.php:255
 Symfony\Component\Console\Command\Command->run() at /var/www/html/magento2/vendor/symfony/console/Application.php:893
 Symfony\Component\Console\Application->doRunCommand() at /var/www/html/magento2/vendor/symfony/console/Application.php:262
 Symfony\Component\Console\Application->doRun() at /var/www/html/magento2/vendor/magento/framework/Console/Cli.php:105
 Magento\Framework\Console\Cli->doRun() at /var/www/html/magento2/vendor/symfony/console/Application.php:145
 Symfony\Component\Console\Application->run() at /var/www/html/magento2/bin/magento:23

maintenance:enable [--ip IP] [--magento-init-params MAGENTO-INIT-PARAMS]
```
3. Exception message isn't saying anything, message not valid at all.

### Additional info
Why it's important to keep `env.php` file read only - on production servers it's the only one way to make sure that no one will disable cache --> website will work fine.

My investigation shown that this issue appeared in https://github.com/magento/magento2/commit/0e373538ded9abe3051d8e9e7a04ca1b6e1d173f.
If i'll remove following observer - it works fine:
https://github.com/magento/magento2/blob/0e373538ded9abe3051d8e9e7a04ca1b6e1d173f/app/code/Magento/PageCache/etc/events.xml#L60-L62

I reviewed this commit and actually it's ok that FPC should not cache pages in maintenance mode, but disabling FPC at all- the wrong way.

What I would suggest - revert changes that were done in https://github.com/magento/magento2/commit/0e373538ded9abe3051d8e9e7a04ca1b6e1d173f and add checking for maintenance flag in when we're sending public `Cache-Control` headers

**Comment by magento-engcom-team** [comment]
:white_check_mark: Confirmed by @engcom-Charlie 
 Thank you for verifying the issue. Based on the provided information internal tickets `MC-19601` were created

**Issue Available**: @engcom-Charlie, _You will be automatically unassigned. Contributors/Maintainers can claim this issue to continue. To reclaim and continue work, reassign the ticket to yourself._

**Comment by irving0mar** [comment]
#cd19latam

**Comment by irving0mar** [comment]
We can't reproduce this issue on Centos 7 and Magento 2.3-develop, give us more information please. For example, what is the OS you are using.

**Comment by ihor-sviziev** [comment]
> We can't reproduce this issue on Centos 7 and Magento 2.3-develop, give us more information please. For example, what is the OS you are using.

I had this issue on Ubuntu 18.04 and on Amazon Linux (Cent OS  based). I don’t think it’s related.
I think main issue - you have to run bin/magento not from root user, because even without permissions root user could write to all files

**Comment by theCapypara** [comment]
#cd-cologne19

**Comment by magento-engcom-team** [comment]
@Parakoopa thank you for joining. Please accept team invitation [here](https://github.com/orgs/magento/invitation) and self-assign the issue.

**Comment by theCapypara** [comment]
This issue is somewhat related to #4955. The best solution for the future would probably to solve that issue, this will also allow changing any cache state with a read-only env.php.

As a simple solution for this issue I will remove the observer ``Magento\PageCache\Observer\SwitchPageCacheOnMaintenance``, which tries to disable the FPC when running maintenance mode to prevent pages being cached. Instead I will change it so, that the PageCache module does not generate cache entries during maintenance mode.

**Comment by VladimirZaets** [comment]
Hi @ihor-sviziev. Thank you for your report. 
The issue has been fixed in magento/magento2#25790 by @Parakoopa in 2.4-develop branch
Related commit(s):
 - [acd42b86af5425d799750f0ceb6c529cd17abba6](https://github.com/magento/magento2/commit/acd42b86af5425d799750f0ceb6c529cd17abba6)
 - [d1b81aa514224b25f1c0fea43400a6ab24e1caeb](https://github.com/magento/magento2/commit/d1b81aa514224b25f1c0fea43400a6ab24e1caeb)

The fix will be available with the upcoming 2.4.0 release.
